src/forward/compiler.jl

```diff
@@ -70,7 +70,91 @@ function dual(ir)
   return ir
 end
 
-@dynamo function _pushforward(_, x...)
+const chainrules_frule_fallback = which(frule, Tuple{Any})
+
+"""
+  has_chain_frule(T)
+
+For a type-tuple `T` e.g. `Tuple{typeof(f), Int, Float64}`, checks if there is a `rrule` defined for it.
+Excluding the generic fallback.
+The first return value is `true` if the `rrule` exists, `false` otherwise.
+If it does not, then the second argument is a list of edges to attach to the CodeInfo for a generated function,
+such that if a suitable rule is defined later, the generated function will recompile.
+"""
+function has_chain_frule(T)
+  m = meta(Tuple{typeof(frule), T.parameters...})
+
+  if m.method !== chainrules_frule_fallback
```

Comment:
I wonder if we should just abstract this to be `has_chainrule(kind, T)`

```suggestion
function has_chain_rule(kind, T)
  m = meta(Tuple{typeof(kind), T.parameters...})

  if m.method !== chainrules_rule_fallback(kind)
```